### DISCUSSION TITLE
Algorithms - do we need this stuff?

### DISCUSSION THREAD

**Discussion by rivantsov** [graphql/graphql-wg#1020] [General]
![UniquenessExample](https://user-images.githubusercontent.com/594969/173213316-a83cbb77-b44c-4869-871a-b049156bd2c3.jpg)

This is '_operation name uniqueness check_' algorithm. Which can be expressed with a simple human-readable sentence - which [actually follows it in the spec](http://spec.graphql.org/October2021/#sec-Operation-Name-Uniqueness.Explanatory-Text): 

> Each named operation definition must be unique within a document when referred to by its name.

Now, do we really need this multi-step, barely readable multi-line nonsense? Would it be enough to simply have this 'humanly' statement? 

Another point of view. We all know **declarative vs imperative dilemma**. Declarative is usually preferable, most of the time. What we do in the spec - we make IMPERATIVE to be our primary way in the spec (we list algorithm first). Then imperative code is followed by DECLARATIVE (this simple sentence). Why?!

Finally, it also feels like a personal insult. I am a server-side developer, and since these checks over schema are executed on the server, it appears its target audience is server developers, just like me. I feel I need to ask - why are you explaining this to me - how to check uniqueness of a string in a list? Your think I am that stupid? and the spec actually says - well, yeah... 

Now seriously, what's the point of this and other 'algorithms'? they take space, cost a lot of effort to write and verify, while bringing ZERO value. The spec gets bloated. I doubt there is anybody who needs these, client or server dev.  

Let's go DECLARATIVE, not imperative. Let's get rid of these.

**Comment by yaacovCR** [comment]
I think the algorithms bring value.

1. The spec PR process is purposefully tightly coupled with the reference implementation. This allows for a rapid and productive feedback cycle for proposed spec changes. Often, a spec change will require re-evaluating after initial implementation, which often raises issues not clear on first impression. Having algorithms within the spec further tightens this feedback cycle, by forcing a pseudo-code implementation which is helpful when planning the graphql-js pr.
2. I don’t think the algorithms are non readable for humans. They are quite readable and should be quite accurate. They are very helpful to implementors and reviewers who are usually human!
3. Some algorithms are so simple that they may seem unnecessary. But then they are pretty short! Because a dividing line between “simple” and “complex” is pretty subjective, it is safer and more practical to insist on uniformity, in my opinion.
4. The spec explicitly states that all equivalent algorithms are compliant, so the imperative portion is non-constraining. If you find them unhelpful, you can always skip them. Since they are helpful to others (even though they may be distracting to those who don’t need them) the most inclusive option is to include them (which does impinge on the readability for those who don’t want them). You can’t please everyone (at least not all of the time).

**Reply by rivantsov** [reply]
Thank you for your thoughtful feedback! Sorry for late reply, had to spend some time reading and thinking

Let me address your argumentation in details. 


> The spec PR process is purposefully tightly coupled with the reference implementation. 

My first question - Is it so tightly coupled? We do have eventual feedback from Ivan that feature X is 'OK' or already even implemented in test branch, ahead of official inclusion in the spec. Is it a 'tight coupling'?

And even if the WG process (!) is coupled with graphql.js, should it affect the spec itself? should the PROCESS of producing the spec affect the content of the spec? In my opinion, the spec should stand on its own, as an informationally complete document, without the need for background context like 'we put this in the spec because that's how we PRODUCE the spec. That's unnecessary connection/dependency to something outside the spec. 

As for the role of reference implementation. I think you are overstating its role, which is mostly utilitarian. Usually reference implementation(s) serve two basic purposes: 
  - a solid proof that feature X CAN BE IMPLEMENTED. This is a proof for us, the **Working Group** , that features are actually implementable. For us here, not for others. For this role, the graphql.js does not even need to be published. Purely utilitarian.
  - serve as a **last-resort source for other implementers** to see how to implement a feature, when an implementer (programmer like me) is stuck, does not know how to implement it (say in c# or java), and wants to 'see the code how' to get a clue. However, this utility is quite limited. graphql.js is a large body of code; JS/ts is a very complex language, and I as an example, have close to zero knowledge of this language, I work in c#. Going to graphql.js for extra info would be absolutely last resort move for me, only after I exhaust all other options. It's not just me, I am sure that many non-js devs are in similar situation. **The utility of graphql.js for supplanting the spec is quite limited.** 
  
  You might say "ah, if you can't use graphql.js code, that's where algorithms in the spec will be helpful". But unfortunately, no. Realistically I do not see any of the algorithms in the spec helping in this case - most are trivial (although quite long sometimes) step by step instructions. Again, a few plain text sentences are much better (and these are there already). May be, just maybe a couple of them (ex: merging fields) - I had to read to understand a bit more, but even this Merging Fields can be better explained with a plain text and a couple of examples of query+output. Surprisingly, there are no merged examples there, not a single one. I truly believe there's a better way explaining it. 
  
  To sum it up - **the spec should stand on its own as a document, without hidden 'implied' dependencies.**. To include algorithms or not should not be because of the process or because there is graphql.js'. The WG process or existence of ref implementation should not be a factor for this decision. 

> 2. I don’t think the algorithms are non readable for humans. They are quite readable and should be quite accurate. 

I do not argue against that, yes, they are readable. The question is 'cost', how much volume they add to the spec, how much effort we, the WG, have to spend accurately writing, verifying and formatting them, and how much extra effort they cause for the reader. I had to spend quite a time and effort reading through every one of them, fearing I can miss something important, and being disappointed each time - no, just trivial stuff. 

Like now, there's a proposal to add algorithm for checking that 'operation type exists'. Absolutely trivial stuff, do you think any implementation will miss this check in the code, or that without the 'algorithm' any implementer would have trouble implementing it? But still, we go for an effort of discussing it, writing it down, approving, merging it.  

**I argue that the added value is very low, while the 'cost' is quite high.** 

> 3. Some algorithms are so simple that they may seem unnecessary. But then they are pretty short! 

The problem is not that they are simple, they are just trivial. Just empty fluff. Checking uniqueness of names. For a first-time reader, the question is - why this trivial stuff is here?!

> 4. The spec explicitly states that all equivalent algorithms are compliant, so the imperative portion is non-constraining. If you find them unhelpful, you can always skip them. 

The problem is that in order to find them unhelpful, **I have to read them!** Carefully, line by line, fearing of missing something important! It takes time, attention and effort. I did that! and I was disappointed every time.

**Reply by yaacovCR** [reply]
> My first question - Is it so tightly coupled?

To advance changes to the spec, a reference implementation PR is required. The spec text must match its algorithms and the algorithms must match the reference implementation. As you say, the process by which the spec is changed is partly encoded within the spec itself! In my opinion, that's a very tight coupling -- but I acknowledge that I have no objective "coupling scale" to which I can refer. I suppose it's tight enough that the above points hold.

> We do have eventual feedback from Ivan that feature X is 'OK' or already even implemented in test branch, ahead of official inclusion in the spec. Is it a 'tight coupling'?

My best read as to what you are getting at with these lines is that there could be certain arguments that the coupling is not so tight. You may be suggesting that there are two in particular:
1. Some features within `graphql-js` may be considered "ok by Ivan" (our core `graphql-js` maintainer): I assume that in this context that you mean that these features do not require a spec change.
2. "Test" branches. We now have PR branches that can be used to test individual spec PR changes, as well as an alpha branch for v17. This process may be somewhat in flux -- it's pretty new after release of v16 and I hope I'm describing it accurately.

Frankly, I think I have misconstrued what you are getting at with these points, because features that don't contradict the spec and the existence of "test" branches do not break the feedback cycle I have described between the spec and the reference implementation and certainly do not change the role of the spec algorithms in bridging the spec text and the reference implementation. Perhaps in some sense, the reference implementation is not as tightly coupled to the spec, but in the sense that matters to this issue, it is still coupled enough for the above points to hold.

> As for the role of reference implementation. I think you are overstating its role, which is mostly utilitarian. Usually reference implementation(s) serve two basic purposes:
> ...

I agree that the "feasibility" and the "reference for other language" reasons for a reference implementation are quite important! But I also think that the clarifying power that an actual implementation brings to a proposed spec __change__ is perhaps even more important! And I find the inclusion of algorithms within the spec itself to a very helpful clarifying force in that regard, really highlighting what has changed. Now, you can't set the stage for how spec changes will change the involved algorithms if you don't specify all the algorithms. So, it all kind of flows together -- or, at least it does for me!

@rivantsov, we have partly heard your perspective as a reader of the spec. I wonder if some sort of tooling within the online spec text to minimize/hide all spec algorithms might be something that you would appreciate.

But, you have also suggested that the inclusion of algorithms inhibits spec writers, and I don't have a similar suggestion on that end. I would say that if it ever so happens that you in the midst of championing a spec change and have new spec text for review, but just don't have the patient for the bothersome algorithm, I would imagine that someone from the community -- perhaps even me -- would be able to donate some time to contributing that text. If you have come up with a `graphql-js` PR for your proposed change, it should be fairly easy to reverse engineer the spec algorithm. And, if the algorithm text generated from your PR differs from your spec text, I suppose we will have our feedback loop right there, even with multiple people originating it. That's about the best practical suggestion I can think of going forward.

**Reply by rivantsov** [reply]
Sorry again for delay in response. 

> To advance changes to the spec, a reference implementation PR is required. 

Agree, no doubt.

> The spec text must match its algorithms and the algorithms must match the reference implementation

The last part 'algorithms must match the ref implementation' ? - where is this coming from? is it stated anywhere? Spec states that implementations may use optimized versions of algorithms in spec. The special thing about graphql.js as reference implementation is that **it is also a real framework widely used by many** . I think the responsibility of illustrating the spec is not their top priority. Conform - yes; verify a new spec feature is doable? yes. Implement algorithms close to spec? I do not think they care about this. They care about performance, usability, maintainability. I doubt they have a single stretch of code matching any spec's algorithm. My own experience, implementing in .NET, field name uniqueness. Do you think I do these steps from spec? No, I use HashSet<string> class to catch the repeated name. Nothing like in spec. Pretty sure similar thing happens in other implementations, including graphql.js.   

So I do not see a valid argument about _tight coupling of spec and graphql.js_ here.  

> Some features within graphql.js may be considered "ok by Ivan" (our core graphql.js maintainer): I assume that in this context that you mean that these features do not require a spec change.

No, nothing like that. A NEW(!) feature proposed in spec, should get 'OK by Graphql.js team' at some early point; and a full implementation, working and tested. It is always about new proposed feature in Spec!


> But I also think that the clarifying power that an actual implementation brings to a proposed spec change is perhaps even more important! 

Well, about clarifying power of implementation - only for some of us. Remember, I do NOT know js/ts enough, so for me graphql.js could as well be graphql.ancient-chinese, same clarifying power. And this is the case for many others coming from .net, java, go, ruby. So goes the clarifying power of graphql.js. 

> And I find the inclusion of algorithms within the spec itself to a very helpful clarifying force in that regard, really highlighting what has changed. 

Including into spec something that _clarifies the change_ (?!). Which change?! The spec, when it is released at some points, has no changes. It is one document, there no changes there! It does not mark sections as added at certain version. It is just one plain document describing ALL features of GraphQL, not broken by individual changes. All this clarifying-change stuff should stay here, in WG discussions, and not go to the final spec. 

> I wonder if some sort of tooling within the online spec text to minimize/hide all spec algorithms might be something that you would appreciate.

I do not think that solves the problem. Spec should be available as downloadable doc, printable. Fancy viewer, OK, that's good. But primary format is plain printable file - which unfortunately contains all this empty fluff in plain sight. But even if you hide it - it would help maybe if I just scan the spec thru on the first read. But then later, if I am serious about GraphQL and going to use it, I have to read it carefully, without skipping anything - just for the fear of missing something important. 

About your offer to help write it. Thank you, but I do not think it solves anything. It is still extra time to explain and discuss with you, and careful read/verification by other members of WG. 

To sum it up. To be honest, I do not find your arguments (through connection to ref implementation and its role) - I do not find them convincing or even working for your case. 

Thank you.

**Reply by yaacovCR** [reply]
Just to clarify one point above, when I wrote that the algorithm in the spec implementation has to match the reference implementation, I meant only in the sense that it must be equivalent, as mentioned by the spec. You are correct that the coupling is inhibited by allowed optimizations, but in my limited experience those optimizations are often later introduced or carefully scrutinized during the review process.

Other than that, I have not much to add, except that I believe the current process is working well.

**Reply by michaelstaib** [reply]
> Well, about clarifying power of implementation - only for some of us. Remember, I do NOT know js/ts enough, so for me graphql.js could as well be graphql.ancient-chinese, same clarifying power. And this is the case for many others coming from .net, java, go, ruby. So goes the clarifying power of graphql.js.

This is not true, I am also a .NET backend engineer but I can read other programming languages. When we initially started on Hot Chocolate we closely used graphql-js and the algorithms. Having these two sources helped immensely clarify things. 

Our todays implementation while much more optimized still reflects in parts the spec algorithms. I personally find it paramount to have both a reference implementation and the algorithms.

However when we discuss new spec features in the working group we are making sure that at least one other non js implementation verifies a spec feature. The ChilliCream team often implements proposed features together with the graphql-js proposals. We closely look at the graphql-js PRs and even give feedback.

**Reply by michaelstaib** [reply]
> Just to clarify one point above, when I wrote that the algorithm in the spec implementation has to match the reference

Its the other way around, the reference implementation needs to follow the spec algorithms.

**Reply by michaelstaib** [reply]
>  there no changes there! It does not mark sections as added at certain version

It does. 

https://github.com/graphql/graphql-spec/blob/main/changelogs/October2021.md

have a look at the diff section.

**Reply by michaelstaib** [reply]
> but I can read other programming languages

@roman do not get me wrong here. I do not mean my comment in a condescending way. I just think that we all are able to read other programming languages. In all honesty I liked the flow/js style more back in the days because it was purer js code. ts often adds more complexity. 

But we all are able to inspect the algorithms in the implementation. True there are maybe constructs that are difficult to understand/js specific but most of the time this can be easily skipped. The essence that you have a debuggable GraphQL reference implementation that shows the spec algorithms in practice is something to cherish.

The first Hot Chocolate parsers was a 1:1 port of the reference implementation. After some time we rewrote it to make it faster and reduce allocations but in essence we were able to almost take the JS code and rewrite it to .NET code.

There also is one aspect about the spec algorithms that is important. The spec allows to implement or come up with better algorithms as long as the outcome is the same. But to understand what the outcome really is in my opinion you need to have the algorithms. With the algorithms integrated in the spec text its so easy to implement a GraphQL server. IMO I enjoy reading the GraphQL spec, its so much more approachable than most specification texts.

**Reply by rivantsov** [reply]
@michaelstaib
just a few comments on your comments (really appreciate it!)

> (my comment) there no changes there! It does not mark sections as added at certain version. (Michael) It does. (link to changelog)

I originally commented on Yaacov's note that 'inclusion of algorithms ... really highlight(s) what has changed' - what changed, where? as if the sections in the spec are marked with 'This is change in 2021 version'. For the reader, the spec in a snapshot of standard at certain version, a complete doc, without any history or marked changes. And of course I know about changelog there. Simply there are no 'changes' in the spec for the reader, it is just a document. The changelog is byproduct of the WG process, and acknowledgement of contributions. Not part of the spec per se. 

> do not get me wrong here. I do not mean my comment in a condescending way. 

No worries, we are all PROs, never take things personally.  

> I just think that we all are able to read other programming languages.

Well, to read the code of a real-world app, you need more than just a language. We code not just in the language, but in the ecosystem. Lots of other things. Language is probably 20% of what you need to learn. Especially true for javascript with its zillion npm packages. I admit, I exaggerated when I said I cannot read graphql.js - I can, to some extent. By my point is - the spec cannot assume this for everybody. And you should not have to learn a new language to learn GraphQL, if you are coming from a different platform. GraphQL is language-neutral. The only 'language' we can freely use/imply is probably JSON - here I would agree everybody knows what it is.

So we cannot rely on graphql.js for extra explanation. Like if you did not understand feature X - go check the ref implementation.  You are backend guy but can read js - lucky you (no sarcasm); but do not imply it for everybody.  
       
>  IMO I enjoy reading the GraphQL spec, its so much more approachable than most specification texts.

No doubt, esthetic aspect in tech is important - good code is esthetically pleasing. Although enjoyment is more personal thing. Reading the spec is more like a really hard work for me - line by line, fearing of missing something important. But when it comes to algorithms - it feels like a prank, I read this trivial stuff, trying to find a meaning, and feels like there's a hidden camera somewhere recording my stupefied face. And somebody is laughing out there.  

Thank you again, I will make another post later, more general thoughts about algorithms

**Comment by rivantsov** [comment]
Hi all, 

I think this thread got side-tracked into discussion about reference implementation (RefImpl), based on an early claim that somehow algorithms in spec are in some direct relation to RefImpl. We started discussing RefImpl's readability, it's role in bringing new features etc. While it is an interesting discussion on its own, I would like to come back to the original subject of algorithms in the spec. 
I would like to reiterate my principal point: 

**The spec is a document that is self-sufficient and it must completely describe the GraphQL protocol** , clearly explaining all the details, enough for successful implementation of servers or clients in any language or platform. That's how the specs work, in general. The question here and now is the use of [didactic](https://www.merriam-webster.com/dictionary/didactic) tools and approaches in the spec for the best readers' experience. 
  
Let's look again at the original example 

![OpNameUniqueness_ext](https://user-images.githubusercontent.com/594969/176053026-3d5b5530-fdcf-4e09-bbab-f34a39a489de.jpg)

It has 3 parts: 

1. Title, the name of the 'feature'
1. Formal Specification - as an algorithm
1. Explanatory text - plain text description of the feature. 

Since the algorithm precedes the Explanatory Text, and is titled 'Formal Specification', then it is the 'official definition/description of the feature, source of truth, and something that one HAS TO READ to fully understand the feature. The third section 'Explanatory Text' is just something extra, optional explanation material that one can skip if he 'got it' from the 'Formal Specification' - which is a multi-step algorithm. 

It is thi front role of the 'algorithms' that I am not happy with. What if it instead was like this: 

1. Title, the name of the 'feature'
1. Plain text description of the feature, preferably short and concise. This is a formal spec of the feature and the real source of truth.  
1. Optional, illustrative, imperative algorithm, but only if it is at least something not trivial and potentially helpful. The reader can skip it if all is clear. 

If it was like this, honestly, I would be OK, and definitely would not start this crusade/discussion. It is the use of the elements, the order, and the role of algorithms as a formal source of truth (vs optional illustrative element) - that's what I am against and suggest to change. And of course use algorithms sparingly, at least skipping the trivial ones like the one shown above. 

I suggest also to leave the Reference implementation alone, and focus on clarity and explanatory power of the spec text itself, without any implied  links to the WG process, Ref implementation, change logs, or anything else. 

Thank you  
Roman

**Reply by yaacovCR** [reply]
I have no objection to change in order.

I also would reiterate my suggestion that a printable format could hides either the text or the algorithms as desired. This is very GraphQL-esque in that you would only get what you ask for.

Alternatively, we could expose the spec itself as a graphql api to achieve the same effect although of course ironically there would be plenty of breaking changes.

**Reply by michaelstaib** [reply]
This is just one of the simplest formal specifications. Essentially to keep it complete.

<img width="754" alt="image" src="https://user-images.githubusercontent.com/9714350/176185991-0d0ba505-06f0-440f-897b-34416581538e.png">

The execution algorithms really is what makes this spec worthwhile. So I disagree.

**Reply by rivantsov** [reply]
Edit: RETRACTED, algorithm is CORRECT, my apologies, keeping the post for history. 

@michaelstaib

It's really ironic you used this algorithm as an example - ironic because it is **INCORRECT**. Not just a typo, but incorrect in a big way. 

Have a look at this section:  [2.9.5 Null Value](https://spec.graphql.org/October2021/#sec-Null-Value). It says that skipping nullable value vs providing null explicitly are not equivalent, and the difference is essential! Quote: _These two forms may be interpreted differently. For example, a mutation representing deleting a field vs not altering a field, respectively._ Which means that the origin of the null value (explicit or from default) should be carried all the way to the resolver. And there is a note there that this applies not only to arguments but also to variables. 

Which means this algorithm cannot return just a _map<varName, varValue>_, it should return something like _map<varName, tuple<nullOriginFlag, varValue>>_. But the algorithm above ignores the null-origin information, simply drops it. Which means that if you implement it, you are not compliant with the spec. So it goes. 

By the way, the same goes for [CoerceArgumentValues](https://spec.graphql.org/October2021/#CoerceArgumentValues()) - another incorrect and misleading algorithm.

This problem is on my TODO list for NGraphQL. I did implement it like the algorithm describes, I had an initial code in my head, double-checked with this CoerceVariableValues, read it couple of times - yep, looks like I got it right, implemented it. Only to discover later this problem. Yes, I admit it now, NGraphQL has a bug, exactly in this place. Of course, it is ultimately my fault, but I put partial blame on this algorithm in the spec. The reason I still did not fix is because I am still thinking how to make this flag available to the resolver, using c# facilities, in an easy and convenient way for resolver coders. 

To sum it up, the way I see it, the problems are two: 
1. The algorithms in the spec are not directly executable, you cannot test them directly, so it is practically impossible to avoid 'bugs' like these. Plus flaws like this become visible only on a big view, when you try to implement the entire spec, not just one algorithm.
1. On a more fundamental level. All algorithms in Validation, Execution sections describe HOW the server should do its job - which is IMHO should not be in the spec. If we go with my suggestion to define: **GraphQL is a communication protocol** (see other discussion), then the spec should be concerned only with _what's on the wire_ : if the request is 'this', then server response is 'that'. How server does it - out of scope for the spec. Therefore, all or most algorithms should be gone.

**Reply by yaacovCR** [reply]
These algorithms are correct and match the spec. The spec text you quote essentially state that users must be able to (A) not send in a value for an argument and (B) send in a null value. The algorithm provides for this by taking different action depending on the _hasValue_ flag in which an entry is only added to the map if this flag is true.

The origin of an actual null value, whether set by user or by a default value within the schema does not matter and need not be recorded. It is only when there is no default value and the user supplies no value that the entry is excluded from the map.

In theory, the spec text you quote could be clarified to note that it is discussing a scenario in which there is no default of null, although that may complicate the discussion there.

This is an example where formal algorithms seem to me to be superior than spec text as explanatory text is by nature often incomplete when attempting to highlight particular points. Having both text and formal algorithms available-- as we currently do -- seems best.

The situation above btw is further slightly confusing in terms of constraining user options with non-nullable input fields. Ie setting an input field to non nullable currently currently both disallows not sending a variable (in the absence of a default value) and disallows sending null. See https://github.com/graphql/graphql-spec/issues/476#issuecomment-401469056 There is no current option to disallow one or the other without both, but supplying a default excludes the possibility of disallowing the former. The algorithm accurately characterizes the current behavior.

**Reply by yaacovCR** [reply]
Corrected typo above at the end of the second paragraph should read "...that the entry is excluded from the map."

**Reply by rivantsov** [reply]
> The algorithm provides for this by taking different action depending on the hasValue flag in which an entry is only added to the map if this flag is true.

Please look at the algorithm again, there is no path where entry is not added. If you have 5 variables defined, you have 5 entries in the output map, never less (or error is thrown). So information is lost - was there null value for variable $x or entry for $x was missing.  

> The spec text you quote essentially state that users must be able to (A) not send in a value for an argument and (B) send in a null value. 

It also says something else - that these two cases are different, and mutation's resolver can take different action depending on was null value provided or not.

**Reply by rivantsov** [reply]
The issue/thread 476 you refer to talks about similar challenge but with Input types - how to express no-value vs null for fields of input types. Similar issue, and it is even more challenging there

**Reply by yaacovCR** [reply]
> Please look at the algorithm again, there is no path where entry is not added.

There is a path. Paraphrasing sections 3g, h, and i we have the following:

g: only add a value to the map when `hasValue` is false when there is a default value
h: only raise a field error if `hasValue` is false if the type is non-nullable
i: add a value if `hasValue` is true

If `hasValue` is false for a nullable type, and no default is set, no value is added to the map, and no field error is raised. (TYPO IN ORIGINAL HAS BEEN CORRECTED SEE BELOW)

> It also says something else - that these two cases are different, and mutation's resolver can take different action depending on was null value provided or not.

They sure can. In JS/TS, this is done by testing to see whether the value is in the map or whether the value is strongly equal to undefined.

**Reply by rivantsov** [reply]
Feels like we are reading different spec 

> If hasValue is false for a non-nullable type, and no default is set, no value is added to the map, and no field error is raised.

from where you deduce this? It is an error. And the case I am talking about is a variable of Nullable type with default null, so this case with non-nullable is irrelevant.

Using 'undefined' in js/ts - good for them, they have a special value for 'missing', other than null (c# does not). But how it relates to this algorithm? does this algorithm mention 'undefined' or any special value?!  - No. It is null, either explicit value, or when entry for variable is missing. And no way for resolver to distinguish.

**Reply by yaacovCR** [reply]
> Feels like we are reading different spec 
>
> > If hasValue is false for a non-nullable type, and no default is set, no value is added to the map, and no field error is raised.
> 

Hi! Sorry I switched non-nullable with nullable there. That's pretty confusing, sorry!

The correct statement is:

> If hasValue is false for a nullable type, and no default is set, no value is added to the map, and no field error is raised.

I edited above

**Reply by yaacovCR** [reply]
> Using 'undefined' in js/ts - good for them, they have a special value for 'missing', other than null (c# does not). But how it relates to this algorithm? does this algorithm mention 'undefined' or any special value?! - No. It is null, either explicit value, or when entry for variable is missing. And no way for resolver to distinguish.

In JS/TS, when no entry exists in the Map, you have the option of (1) explicitly checking to see whether the key is in the map or (2) checking to see if the value is undefined. I mentioned both of the possibilities because I wanted to make clear that in other languages, you may have only a single choice, i.e. (1).

Of course, I forgot to make that clear.

**Reply by yaacovCR** [reply]
> the case I am talking about is a variable of Nullable type with default null

A nullable type with default of null is assigned a value of Null. The spec section you quote where it resolvers must be allowed to distinguish between cases where the argument is omitted and where it is explicitly set to null by the user is referring to a case does not include a default of null. This can be checked by following the algorithm.

3/g/i says a value is added to the map if `hasValue` is false and there is a default, including null. If there is no default, the value is not added to the map.
3/h describes non-nullable types, not relevant to this scenario.
3/i describes `hasValue` of true, such that if the value is null, it is added to the map.

So without a default, with a nullable type, there are two different outcomes if the user specifies null, 3/i says that it is added to the map, whereas if no value is provided, none of the three "action" portions of the algorithm pertains, and no item is added to the map.

Resolver code can distinguish between whether `null` exists in the map for the variable, or if there is no entry for the variable.

I think the ambiguity to me seems to be that you believe that the resolver should still be able to tell whether a default of null was used or not, and I don't think that's the case.

**Reply by benjie** [reply]
Indeed:

- it is irrelevant if a null originates from a `defaultValue` or from the user, either way it is `null`
- GraphQL distinguishes between a value not being present (neither provided by the user, nor by a default value) versus a value being null (either explicitly from the user, or implicitly from a default value)

**Reply by rivantsov** [reply]
ok, now I'm getting somewhere and starting to understand something. Let me spend some more time to digest these details. Thanks guys!

**Reply by rivantsov** [reply]
OK, I retract, these algorithms are correct, at least for missing values tracking. My apologies for starting this messy thread. Thank you. 

PS the algorithms are correct, but are not easy to go through for particular scenarios. A few explanatory sentences (like @benjie 's comment above, and a note that coerced args/vars map count might be less than 'definition set') - would have helped a lot.

**Reply by benjie** [reply]
This is one of the reasons algorithms are core to the spec - they help to convey this kind of subtle information with minimal ambiguity. Care must be taken when interpreting the algorithms though, which is why the spec also includes explanatory text; if the explanatory text here is insufficient as you suggest then a good course of action would be to raise a PR to improve the explanation.

**Comment by leebyron** [comment]
I always welcome a spicy take, but I must say I strongly disagree with this one. I find the algorithms as one of the primary ways the spec brings value.








> Do we really need this multi-step, barely readable multi-line nonsense? Would it be enough to simply have this 'humanly' statement?

English phrases are notoriously ambiguous. There may be cases where we have simple algorithms which could be stated in English. It's quite useful to have _both_ actually. English should be easier to read and understand intent where algorithms should disambiguate and be very clear as to the explicit order of operations.

We also have algorithms which are considerably more complicated. It would be a huge problem to omit complex algorithms in favor of ambiguous English prose. Rather than having some rule where only complicated algorithms get written in full form, it's useful to describe all algorithms uniformly.

> Another point of view. We all know declarative vs imperative dilemma. Declarative is usually preferable, most of the time. What we do in the spec - we make IMPERATIVE to be our primary way in the spec (we list algorithm first). Then imperative code is followed by DECLARATIVE (this simple sentence). Why?!

This is good feedback, and a case where we are not always consistent across the spec document. I like the principle that you should be able to read the spec text top to bottom linearly without encountering knowledge large gaps along the way. There are cases where we aid this by explaining a concept at a high level relative to some section if it's described in detail further below.

Algorithms appearing before prose which explains their purpose and high level behavior is in fact backwards according to that principle.

> It also feels like a personal insult. I am a server-side developer, and since these checks over schema are executed on the server, it appears its target audience is server developers, just like me. I feel I need to ask - why are you explaining this to me - how to check uniqueness of a string in a list? Your think I am that stupid? and the spec actually says - well, yeah...

I think it's worth pointing out here that the purpose of the algorithms is not to tell server-side developers how to write a program. I agree with you that developers are perfectly capable of writing their own code, and in fact [we encourage that](https://spec.graphql.org/draft/#sel-GABDLFABABjK1oE) in the spec.

Algos should disambiguate and explain to all developers but most importantly _consumers_ of a GraphQL API what behavior they should expect. If a behavior is not observable to a client, then it may not be required to be described in a spec.

> Now seriously, what's the point of this and other 'algorithms'? they take space, cost a lot of effort to write and verify, while bringing ZERO value. 

Over time I have seen a very large number of examples where explicit algorithms have helped disambiguate and help resolve inconsistencies in the ecosystem, help guide spec discussions, discussions around building tools, and countless other cases where a lack of clarity would have resulted in real cost but clearly defined algos avoided that outcome (or we faced real cost, where writing an algorithm more clearly helped resolve it).

So I have to firmly disagree with the statement that they bring zero value. In fact I've experienced their great value first hand.

**Reply by rivantsov** [reply]
@leebyron 
Thank you for your detailed response. I would be honest, I am not convinced, but that's ok. We exchanged opinions that's the main point, and I think we can now put the issue to rest. I will leave the discussion open for a few days so others can find and read your response, and then close it (wonder if its possible). 
Thank you again
